### DISCUSSION TITLE
Viterbi function with multiple imputation

### DISCUSSION THREAD

**Discussion by sarahchinn** [bmcclintock/momentuHMM#143] [Q&A] [answered]
I am using CRAWL to regularize my polar bear tracks for a location every 4 hr. I imputed 20 tracks to account for uncertainty and ran the MIfitHMM function for the 20 tracks (MI_mod). I would like to assign the most likely sequence of states (n=2) for each location using the Viterbi algorithm function (viterbi(MI_mod)) but that function only takes momentuHMM or momentHierHMM objects. How do I accomplish with a miHMM object, given by the multiple imputation method?

I have provided data for 1 bear and my code for fitting the MI_mod below:
[pb_21678_MIdata_5.18.22.csv](https://github.com/bmcclintock/momentuHMM/files/8757271/pb_21678_MIdata_5.18.22.csv)

 # folder where multiple imputation (from CRAWL) data is located
indir<-'C:\\Users\\....'

csvlist<-c(list.files(indir,pattern = ".csv$",ignore.case=T))

for(i in csvlist){
  raw<-read.csv(paste(indir,i,sep='/'))
  
  # Convert times to POSIX
  raw$date <- as.POSIXct(raw$datetime, format = "%Y-%m-%d %H:%M:%S")
  
  # Create list of imputed tracks
  ind_imp <- c(0, which(raw$datetime[-1] < raw$datetime[-nrow(raw)]), nrow(raw))
  data_list <- list()
  raw$imp <- NA
  for(i in 1:(length(ind_imp) - 1)) {
    subdata <- prepData(raw[(ind_imp[i] + 1):ind_imp[i + 1],], 
                        coordNames = c("X", "Y"))
    data_list[[i]] <- subdata
    raw$imp[(ind_imp[i] + 1):ind_imp[i + 1]] <- i
  }
  
  # Fit model to multiple imputation data
    MI_mod <- MIfitHMM(miData = data_list, nbStates = 2,
                     Par0 = list(step = c(100, 300, 100, 300),
                                 angle = c(pi, pi, 0.5, 0.1)),
                     dist = list(step = "gamma", angle = 'wrpcauchy'),
                     estAngleMean = list(angle = TRUE))
  
  # save the model output summary
  sink("MI_mod_21678_5.22.txt")
  print(MI_mod)
  sink()
  
}

MI_mod
plot(MI_mod)

 # Assign the probable sequence of states with the Viterbi algorithm
viterbi(MI_mod)
**Error here because MI_mod is a miHMM object but viterbi() requires a momentHMM or momentHierHMM object

**Comment by bmcclintock** [comment, marked as answer]
When `poolEstimates=TRUE` then the `MIpool` function is called after the model fits. The output of `MIpool` provides the most frequent state assignment for each time step based on the Viterbi algorithm for each individual model fit and is accessible via 
```
MI_mod$miSum$Par$states
```
See `?MIfitHMM`, `?MIpool`, and the {momentuHMM} [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf) for further details (particularly Section 2.8, references therein) and examples.
